internal/nix/shell.go

```diff
@@ -70,14 +71,63 @@ type Shell struct {
 
 type ShellOption func(*Shell)
 
-// DetectShell attempts to determine the user's default shell.
-func DetectShell(opts ...ShellOption) (*Shell, error) {
-	path := os.Getenv("SHELL")
-	if path == "" {
-		return nil, errors.New("unable to detect the current shell")
+// NewShell initializes the Shell struct so it can be used to start a shell environment
+// for the devbox project.
+func NewShell(opts ...ShellOption) (*Shell, error) {
+	shPath, err := shellPath()
+	if err != nil {
+		return nil, err
+	}
+	sh := initShellBinaryFields(shPath)
+
+	for _, opt := range opts {
+		opt(sh)
 	}
 
-	sh := &Shell{binPath: filepath.Clean(path)}
+	debug.Log("Recognized shell as: %s", sh.binPath)
+	debug.Log("Looking for user's shell init file at: %s", sh.userShellrcPath)
+	return sh, nil
+}
+
+// shellPath returns the path to a shell binary, or error if none found.
+func shellPath() (path string, err error) {
+	defer func() {
+		if err != nil {
+			path = filepath.Clean(path)
+		}
+	}()
+
+	// First, check the SHELL environment variable.
+	path = os.Getenv("SHELL")
+	if path != "" {
+		debug.Log("Using SHELL env var for shell binary path: %s\n", path)
+		return path, nil
+	}
+
+	// Second, attempt to find the path to one of these common shells
+	shells := []string{"sh", "bash", "zsh", "ksh", "fish", "dash", "ash"}
```

Comment:
Instead of searching in PATH what if we just use Nix's bash shell as the fallback? Otherwise we might find `/bin/sh` which could be something like dash, which isn't great for interactive shells.

Reply:
how does one find the path to `nix`'s bash shell? 

I'd considered this approach initially. 

The problem I ran into is that the straightforward approach for this would be to do `nix develop`. But that needs other processing to make work properly. This would diverge logic from the regular implementation, which is what I was seeking to avoid.

Alternatively, we can get the path to bash in the `/nix/store` and use that here. Not fully sure how to do that, but I can look into it.

Reply:
I think I found it: `nix eval --raw nixpkgs/<commit>#bash`

```
❯ /nix/var/nix/profiles/default/bin/nix eval --raw nixpkgs/52e3e80afff4b16ccb7c52e9f0f5220552f03d04#bash
/nix/store/1bsjl5incfnszv7scdh4d02sh45vw2w1-bash-5.1-p16
```
and to confirm:
```
❯ /nix/var/nix/profiles/default/bin/nix develop .devbox/gen/flake/flake.nix
path '/Users/savil/code/jetpack/devbox/examples/testdata/go/go-1.19/.devbox/gen/flake/flake.nix' does not contain a 'flake.nix', searching up
warning: Git tree '/Users/savil/code/jetpack/devbox/examples/testdata/go/go-1.19/.devbox/gen/flake' is dirty

Savil-Srivastavas-MacBook-Pro:go-1.19 savil$ which bash
/nix/store/1bsjl5incfnszv7scdh4d02sh45vw2w1-bash-5.1-p16/bin/bash

Savil-Srivastavas-MacBook-Pro:go-1.19 savil$ echo $SHELL
/nix/store/1bsjl5incfnszv7scdh4d02sh45vw2w1-bash-5.1-p16/bin/bash
```

Reply:
oh need to find the non-flakes way to do this too...

Reply:
for non flakes:
```
❯ /nix/var/nix/profiles/default/bin/nix-instantiate --eval --strict --json --expr 'let
pkgs = import (fetchTarball { url = "https://github.com/nixos/nixpkgs/archive/52e3e80afff4b16ccb7c52e9f0f5220552f03d04.tar.gz"; })
{};
in {inherit(pkgs.bash) outPath; }'
"/nix/store/1bsjl5incfnszv7scdh4d02sh45vw2w1-bash-5.1-p16"
```

Reply:
useful command to get the attributes of a package:
```
/nix/var/nix/profiles/default/bin/nix-instantiate --eval --strict --json --expr 'let
pkgs = import (fetchTarball { url = "https://github.com/nixos/nixpkgs/archive/52e3e80afff4b16ccb7c52e9f0f5220552f03d04.tar.gz"; })
{};
in pkgs.lib.attrNames pkgs.bash'
```